### ISSUE TITLE
No error when 2 materials with the same names

### ISSUE THREAD

**Issue by xmik** [tomzo/gocd-yaml-config-plugin#3]
My pipeline depends on 2 upstream pipelines (P1 and P2) and 1 git repository. My config contains:

``` yaml
materials:
      git:
        git: "git@<path_to_git_project>.git"
        branch: ci
      upstream:
        pipeline: P1
        stage: test
      upstream:
        pipeline: P2
        stage: itest
```

It would be nice if there was an error saying that I configured 2 materials with the same name: `upstream`. Currently such configuration results in ignoring 1 upstream pipeline material.

**Comment by sanjusoftware** [comment]
the YAML file is read using com.esotericsoftware.yamlbeans.YamlReader in this plugin...  it automatically will override the value for a keyword with the last value found in the yaml. so e.g last material definition of 'upstream' in this issue. So, it'll load following:

```yaml
      upstream:
        pipeline: P2
        stage: itest
```

@tomzo I am thinking to fix this by introducing a YAML validator, but wanted to get your opinion as well on the approach. How do you think we can best handle this?

**Comment by tomzo** [comment]
As you have noted, the problem is that after yamlbeans has loaded yaml into a map [here](https://github.com/tomzo/gocd-yaml-config-plugin/blob/master/src/main/java/cd/go/plugin/config/yaml/YamlFileParser.java#L26) there is no way to tell that there was a duplicate key in yaml file. `com.esotericsoftware.yamlbeans.YamlReader ` looses that information.
I think the correct way to fix this is by overriding [this method](https://github.com/EsotericSoftware/yamlbeans/blob/master/src/com/esotericsoftware/yamlbeans/YamlReader.java#L112) in `YamlReader` or even copy entire YamlReader class to source here and modify as needed. Essentially we want to ensure that it does not override the value for a keyword with the last value found in the yaml. Such fix would take care of a few other places where similar errors can be make: `pipelines:`, `jobs:`, `:tabs`, etc.- all these require unique keys.
That was my idea so far, but I think YAML validator is a good approach too, I guess it has more applications then, e.g. to validate keywords.

**Comment by sanjusoftware** [comment]
how about using the [yamllint](https://github.com/stma/yamllint-maven-plugin) as the first step [here](https://github.com/tomzo/gocd-yaml-config-plugin/blob/master/src/main/java/cd/go/plugin/config/yaml/YamlFileParser.java#L22). ? yamllint gives error on finding duplicate keys in a yaml and we can throw that error out. we will have add yamllint as a gradle dependency. What I have not yet been able to figure out yet though, is how to use call its validator from the java program :(

**Comment by tomzo** [comment]
I don't think that the maven plugin you pointed actually uses the python yamllint. It is actually based on snakeyaml, see [this dependency](https://github.com/stma/yamllint-maven-plugin/blob/master/pom.xml#L66). I don't think it is worth going outside of java just to implement validation. 
If snakeyaml fails on duplicate keys (I'm not sure if it does), we can also change parser to [snakeyaml](https://bitbucket.org/asomov/snakeyaml/overview), but that is quite some work.

**Comment by sanjusoftware** [comment]
I raised a similiar question on the [yamlbeans google fourm](https://groups.google.com/forum/#!topic/yamlbeans-users/dZKp3LnttOM). does this makes any sense  to you @tomzo ? I think we are close, but we'l have to override YAMLReader methods!

**Comment by tomzo** [comment]
It seems right. Although the [method mentioned by Nate](https://github.com/EsotericSoftware/yamlbeans/blob/master/src/com/esotericsoftware/yamlbeans/YamlReader.java#L197) is private so I guess it has to be duplicated here.

I think we can add a class `ValidatingYamlReader` inheriting `YamlReader` with  `readValue` overriden
```java
@Override
protected Object readValue (Class type, Class elementType, Class defaultType) throws YamlException, ParserException,
		TokenizerException {
		
// same as in original
//...
		return readValueInternalNoDuplicates(this.chooseType(tag,  defaultType, type), elementType, anchor);
	}
```
And then copy implementation of `readValueInternal` to a new method `readValueInternalNoDuplicates` and implement unique key checking.

Another option would be to contribute to yamlbeans and add there an option for validating duplicate keys.

**Comment by sanjusoftware** [comment]
yea, I asked Nate if he thinks that its a right feature to be added to the YamlBeans itself, then I can contribute there... 

after reading the code, I think we'll have to declare a set of keys just before the while loop beings [here](https://github.com/EsotericSoftware/yamlbeans/blob/master/src/com/esotericsoftware/yamlbeans/YamlReader.java#L311) and will have add a check just before this line [here] (https://github.com/EsotericSoftware/yamlbeans/blob/master/src/com/esotericsoftware/yamlbeans/YamlReader.java#L325), where we know what key is being added to the current HashMap object. what do you think @tomzo ?

**Comment by tomzo** [comment]
It sounds about right. 
I would start by adding a test which reproduces duplicate-key case, debug the code to see where the value is replaced, then I would attempt at making changes in `readValueInternal`.

**Comment by sanjusoftware** [comment]
👍 absolutely. let me take a stab at that!

**Comment by tomzo** [comment]
Great, thanks for signing up on this :)

**Comment by sanjusoftware** [comment]
@tomzo because of the difficulties as explained [here](https://groups.google.com/forum/#!topic/yamlbeans-users/dZKp3LnttOM) in extending the YamlReader class as we decided above, what I ended up doing is raising a [pull request](https://github.com/EsotericSoftware/yamlbeans/pull/58) to [yamlbeans](https://github.com/EsotericSoftware/yamlbeans). I hope this gets accepted. Once that is done, we can rather change the usage of YamlReader [here](https://github.com/tomzo/gocd-yaml-config-plugin/blob/master/src/main/java/cd/go/plugin/config/yaml/YamlFileParser.java#L25) to something like below:

```java
YamlReader reader = new YamlReader(new InputStreamReader(inputStream), false);
// the second parameter is the addition which I did which is boolean flag to set allowDuplicates. in our case it'll be false and we'll get YamlException which we can catch and throw back appropriately and we should be good
```
I'll add appropriate tests to support the same. Let me know if you see any concerns or think differently.

**Comment by tomzo** [comment]
I skimmed through the PR. Looks very good. 
Calling yaml reader with just one option seems like much cleaner solution than copy-pasting methods here. :+1:

**Comment by sanjusoftware** [comment]
@tomzo kind of got stuck here at the last step... my [PR]((https://github.com/EsotericSoftware/yamlbeans/pull/58)) shows that the whole file changed because of line endings etc.. I tried correcting it, but not sure where I am going wrong... do you have any pointers how to fix it!!

**Comment by tomzo** [comment]
@sanjusoftware try some of [these](http://stackoverflow.com/questions/16768776/convert-line-endings) to add carriage returns.

**Comment by sanjusoftware** [comment]
@tomzo I am not able to move forward.. really stuck..not able to convert the files in PR back to how they should have been!! would be great if you could help :(

**Comment by sanjusoftware** [comment]
@tomzo finally, our change to the yamlbeans is [here](https://github.com/EsotericSoftware/yamlbeans#duplicate-key-validation). waiting for @NathanSweet to release a new version of the jar :) 👍

**Comment by sanjusoftware** [comment]
should be good to close this one now! 👍

**Comment by tomzo** [comment]
Fixed by #17, released in 0.3.0
